pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	if blockID == "" {
+		http.Error(w, "missing block ID", http.StatusBadRequest)
+		return
+	}
+
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
+	if err != nil {
+		http.Error(w, "invalid tenant ID", http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "received request to complete block upload", "user", tenantID,
+		"block_id", blockID, "content_length", r.ContentLength)
+
+	dec := json.NewDecoder(r.Body)
+	var meta metadata.Meta
+	if err := dec.Decode(&meta); err != nil {
+		http.Error(w, "malformed request body", http.StatusBadRequest)
+		return
+	}
+
+	bkt := bucket.NewUserBucketClient(tenantID, c.bucketClient, c.cfgProvider)
+
+	exists := false
+	if err := bkt.Iter(ctx, blockID, func(pth string) error {
+		exists = strings.HasSuffix(pth, ".lock")
+		return nil
+	}); err != nil {
+		level.Error(c.logger).Log("msg", "failed to iterate over block files", "user", tenantID,
+			"block_id", blockID, "err", err)
+		http.Error(w, "failed iterating over block files in object storage", http.StatusBadGateway)
+		return
+	}
+	if !exists {
+		level.Debug(c.logger).Log("msg", "no lock file exists for block in object storage, refusing to complete block",
+			"user", tenantID, "block_id", blockID)
+		http.Error(w, "block upload has not yet been initiated", http.StatusBadRequest)
+		return
+	}
+
+	level.Debug(c.logger).Log("msg", "completing block upload", "user",
+		tenantID, "block_id", blockID, "files", len(meta.Thanos.Files))
+
+	if err := c.sanitizeMeta(blockID, tenantID, &meta); err != nil {
+		level.Error(c.logger).Log("msg", "failed to upload meta.json", "user", tenantID,
+			"block_id", blockID, "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	if err := c.validateBlock(ctx, w, blockID, tenantID, bkt, meta); err != nil {
+		level.Error(c.logger).Log("msg", "failed validating block", "user", tenantID,
+			"block_id", blockID, "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	// Upload meta.json so block is considered complete
+	if err := c.uploadMeta(ctx, w, meta, blockID, tenantID, bkt); err != nil {
+		var eBadGW errBadGateway
+		if errors.As(err, &eBadGW) {
+			level.Error(c.logger).Log("msg", eBadGW.message, "user", tenantID,
+				"block_id", blockID, "err", err)
+			http.Error(w, eBadGW.message, http.StatusBadGateway)
+			return
+		}
+		var eBadReq errBadRequest
+		if errors.As(err, &eBadReq) {
+			level.Warn(c.logger).Log("msg", eBadReq.message, "user", tenantID,
+				"block_id", blockID)
+			http.Error(w, eBadReq.message, http.StatusBadRequest)
+			return
+		}
+
+		level.Error(c.logger).Log("msg", "failed to upload meta.json", "user", tenantID,
+			"block_id", blockID, "err", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
```

Comment:
What is the point of distinguishing these errors? We can simply return internal error if anything fails at this stage.

Reply:
Fixed.